torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -75,10 +75,193 @@ def powerSGD_hook(
     bucket,
 ) -> torch.futures.Future:
     """
-    This DDP communication hook implements a simplified PowerSGD gradient compression
+    This DDP communication hook implements the original PowerSGD gradient compression
     algorithm described in https://arxiv.org/abs/1905.13727.
     Once gradient tensors are aggregated across all workers, this hook applies
     compression as follows:
+    1) Views the input flattened 1D gradient tensor as two groups of per-parameter tensors:
+    high-rank tensors and vector-like rank-1 tensors (for biases).
+    2) Handles rank-1 tensors by allreducing them without compression:
+        2.1) Allocate contiguous memory for those rank-1 tensors,
+        and allreduces all the rank-1 tensors as a batch, without compression;
+        2.2) Copies the indvidual rank-1 tensors from the contiguous memory back to the input tensor.
+    3) Handles high-rank tensors by PowerSGD compression:
+        3.1) For each high-rank tensor M, creates two low-rank tensors P and Q for decomposing M,
+        such that M = PQ^T, where Q is initialized from a standard normal distribution and orthogonalized;
+        3.2) Computes each P in Ps, which is equal to MQ;
```

Comment:
How do we guarantee that Q = (Q^T)-1? I'm guessing this comes from the orthogonality of the matrix?

Reply:
Exactly. This is basically the [definition of orthogonal matrix](https://en.wikipedia.org/wiki/Orthogonal_matrix).